src/peft/tuners/xlora/model.py

```diff
@@ -0,0 +1,323 @@
+from typing import Any, Dict, List, Optional, Union
+
+import torch
+import torch.nn as nn
+from safetensors.torch import save_model  # type: ignore
+from transformers import PreTrainedModel
+
+from peft.tuners.lora.model import LoraModel
+from peft.utils.peft_types import PeftType
+
+from .. import lora
+from .classifier import InhibitorFlagPayload, Number, XLoraClassifier
+from .config import XLoraConfig
+from .insertion import XLoraLayer
+
+
+def convert_layers_to_xlora(
+    base: nn.Module,  # PeftModel
+    config: XLoraConfig,
+) -> int:
+    """
+    Returns the number of swapped layers.
+    """
+    total_swapped = 0
+
+    for module in base.modules():
+        if isinstance(module, lora.LoraLayer):
+            new_layer = XLoraLayer(
+                model=base,
+                target_forward=module.forward,
+                layer_number=total_swapped,
+                config=config,
+            )
+            module.forward = new_layer.forward  # type: ignore[method-assign]
+            total_swapped += 1
+
+    return total_swapped
+
+
+class XLoraModel(LoraModel):
+    """
+    Creates an X-LoRA (Mixture of LoRA experts), model from a pretrained transformers model.
+
+    The method is described in detail in https://arxiv.org/abs/2402.07148.
+
+    Args:
+        model ([`torch.nn.Module`]): The model to be adapted.
+        config ([`XLoraConfig`]): The configuration of the Lora model.
+        adapter_name (`str`): The name of the adapter, does not affect the LoRA adapter names.
+
+    Returns:
+        `torch.nn.Module`: The X-LoRA model.
+
+    Example:
+        ```py
+        >>> from transformers import AutoModelForCausalLM, AutoConfig
+        >>> from peft import LoraConfig, PeftModel, get_peft_model, prepare_model_for_int8_training
+
+        >>> model_config = AutoConfig.from_pretrained("mistralai/Mistral-7B-Instruct-v0.1")
+        >>> config = XLoraConfig(
+        ...     task_type="CAUSAL_LM",
+        ...     hidden_size=model_config.hidden_size,
+        ...     xlora_depth=4,
+        ...     adapters={
+        ...         "adapter_1": "./path/to/the/checkpoint/",
+        ...         "adapter_2": "./path/to/the/checkpoint/",
+        ...         "adapter_n": "./path/to/the/checkpoint/",
+        ...     },
+        ... )
+
+        >>> model = AutoModelForCausalLM.from_pretrained(
+        ...     "mistralai/Mistral-7B-Instruct-v0.1",
+        ...     trust_remote_code=True,
+        ...     use_flash_attention_2=False,
+        ...     device_map="cuda:0",
+        ...     torch_dtype=torch.bfloat16,
+        ... )
+        >>> model = prepare_model_for_int8_training(model)
+        >>> xlora_model = get_peft_model(model, config)
+        ```
+    """
+
+    def __init__(
+        self,
+        model: nn.Module,
+        config: Union[dict[str, XLoraConfig], XLoraConfig],
+        adapter_name: str,
+    ) -> None:
+        super().__init__(model, config, adapter_name)
+
+    def _xlora_post_init(
```

Comment:
I'm not sure. Most of those are deeply nested by the time `post_init_lora` is called, so I thought it would increase readability to pass it this way. Would you prefer it if they are accessed through the PeftModel?